### PR TITLE
Chore: Types & utils improvements

### PR DESCRIPTION
**Pull request by lukasaric**
## Motivation and Resolution

- introduce `typed` util, which groups `typeof` like values checks (docs needs to be updated)
- improve types overall (return types, interfaces etc)
- remove `url` util
- add `assert.test.ts`

It resolves: #1214

## Usage related changes

_n.a_

## Development related changes

_n.a_

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Linked the issues which this PR resolves
- [ ] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

File: src/utils/typed.ts
Comment (penovicp) [inline src/utils/typed.ts:1]: I've been meaning to do a refactoring where a similar file would be introduced so the basic `typeof` wrapper functions could be aggregated from `utils/num` and `utils/shortString`. Seems like this is a convenient PR to include that change.

Comment (lukasaric) [reply src/utils/typed.ts:1]: yeap, something that I also had in my mind, that's why this one is still in draft and I did not updated the description because of the potential TODO's :)

Comment (lukasaric) [reply src/utils/typed.ts:16]: @penovicp, what's the process for introducing new versioned docs? I'm asking since we'll need to update the versioned docs since utilities are changed now

Comment (penovicp) [reply src/utils/typed.ts:16]: The active pre-release documentation is available as `Next` in the dropdown, the `guides` are written manually and tracked by version control and the `API` segment is generated automatically. The versioned docs are essentially snapshots of the `Next` documentation when a starknet.js version is promoted to a full release (merged to `main`), usually there is no need change them. So the changes in this PR will be automatically available as `Next` once it is merged to develop.

And just to provide the full context, the customary approach for generating new versioned docs is for one of the maintainers to run the `[Manual] Documentation Version PR` workflow on the core repo that generates the docs and opens a corresponding PR. This can be tested on any GitHub fork or the workflow script steps can be reproduced to check the behaviour locally.

File: src/utils/encode.ts
Comment (penovicp) [inline src/utils/encode.ts:272]: "without" typo

File: src/utils/stark.ts
Comment (penovicp) [inline src/utils/stark.ts:292]: `{TransactionVersion}` doesn't match the `ETransactionVersion | undefined` return type from the function signature (although I'm pretty sure it doesn't matter and the `@returns` types are mostly unused).

File: src/utils/url.ts
Comment (penovicp) [inline src/utils/url.ts:1]: It should be possible to remove this dependency now.

Comment (lukasaric) [reply src/utils/url.ts:1]: nice catch 🎯

**Review by penovicp — APPROVED** [review]

File: src/utils/fetchPonyfill.ts
Comment (ivpavici) [inline src/utils/fetchPonyfill.ts:9]:
```suggestion
  (!isUndefined(global) && makeFetchCookie(global.fetch)) || // use builtin fetch in node, react-native and service worker if available
```

Comment (ivpavici) [inline src/utils/fetchPonyfill.ts:9]:
```suggestion
  (!isUndefined(global) && makeFetchCookie(global.fetch)) || // use builtin fetch in node, react-native and service worker if available
```

File: src/utils/encode.ts
Comment (ivpavici) [inline src/utils/encode.ts:190]:
```suggestion
 * @example
```

I think examples are without this extra *

**Review by ivpavici — APPROVED** [review]

**Pull request merged**
